lib/webmention/client.rb

```diff
@@ -1,95 +1,47 @@
 
-    # Public: Returns an array of links contained within the url.
-    attr_reader :links
+      @source = source
 
-    # Public: Create a new client
-    #
-    # url - The url you want us to crawl.
-    def initialize(url)
-      @url = URI.parse(url)
-      @links ||= Set.new
-
-      raise ArgumentError, "`#{@url}` is not a valid URL." unless Webmention::Client.valid_http_url?(@url)
+      raise ArgumentError, 'source must be an absolute URL (e.g. https://example.com)' unless source_uri.absolute?
     end
 
-    # Public: Crawl the url this client was initialized with.
-    #
-    # Returns the number of links found.
-    def crawl
-      @links ||= Set.new
-
-      raise ArgumentError, 'URL cannot be nil.' if @url.nil?
-
-      Nokogiri::HTML(open(url)).css('.h-entry a').each do |link|
-        link = link.attribute('href').to_s
-
-        @links.add(link) if Webmention::Client.valid_http_url?(link)
-      end
-
-      @links.count
+    def send_all_mentions
+      mentioned_urls.each_with_object({}) { |url, hash| hash[url] = send_mention(url) }
     end
 
-    # Public: Sends mentions to each of the links found in the page.
-    #
-    # Returns the number of links mentioned.
-    def send_mentions
-      crawl if links.nil? || links.empty?
+    def mentioned_urls
+      raise UnsupportedMimeTypeError, "Unsupported MIME Type: #{source_response.mime_type}" unless parser_for_mime_type
 
-      sent_mentions_count = 0
-
-      links.each do |link|
-        endpoint = Webmention::Endpoint.discover(link)
+      @mentioned_urls ||= parser_for_mime_type.new(source_response).results
+    end
 
-        next unless endpoint
+    def send_mention(target)
+      endpoint = IndieWeb::Endpoints.get(target).webmention
 
-        sent_mentions_count += 1 if Webmention::Client.send_mention(endpoint, url, link)
-      end
+      return unless endpoint
 
-      sent_mentions_count
+      HttpRequest.post(Addressable::URI.parse(endpoint), source: @source, target: target)
+    rescue IndieWeb::Endpoints::IndieWebEndpointsError => exception
+      raise Webmention.const_get(exception.class.name.split('::').last), exception
     end
 
-    # Public: Send a mention to an endoint about a link from a link.
-    #
-    # endpoint - URL to send mention to.
-    # source - Source of mention (your page).
-    # target - The link that was mentioned in the source page.
-    #
-    # Returns a boolean or the response if full_response == true.
-    def self.send_mention(endpoint, source, target, full_response = false)
-      response = HTTParty.post(absolute_endpoint(endpoint, target), body: { source: source, target: target })
+    private
 
-      return response if full_response
-      return response.code == 200 || response.code == 202
-    rescue
-      return false
+    def parser_for_mime_type
+      @parser_for_mime_type ||= Parsers.registered[source_response.mime_type]
     end
```

Comment:
Currently only supports parsing HTML pages, but looking to add JSON and other relevant MIME types in a future update.